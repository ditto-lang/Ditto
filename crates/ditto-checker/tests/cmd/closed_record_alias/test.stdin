module Test exports (..);

type alias Props = { a: Int, b: Bool, c: Unit };

get_a = fn (props: Props) -> props.a;

five = get_a({ a = 5, b = true, c = unit });
