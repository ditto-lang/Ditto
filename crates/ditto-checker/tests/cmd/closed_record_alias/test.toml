bin.name = "ditto-checker-testbin-check-module"
args = ["closed_record_alias.ditto"]
fs.sandbox = true
