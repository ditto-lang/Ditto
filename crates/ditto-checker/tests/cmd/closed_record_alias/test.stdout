{
  "module_name": [
    "Test"
  ],
  "exports": {
    "types": {
      "Props": {
        "Alias": {
          "doc_comments": [],
          "doc_position": 0,
          "kind": "Type",
          "aliased_type": {
            "type": "RecordClosed",
            "data": {
              "kind": "Type",
              "row": {
                "a": {
                  "type": "PrimConstructor",
                  "data": "Int"
                },
                "b": {
                  "type": "PrimConstructor",
                  "data": "Bool"
                },
                "c": {
                  "type": "PrimConstructor",
                  "data": "Unit"
                }
              }
            }
          },
          "alias_variables": []
        }
      }
    },
    "constructors": {},
    "values": {
      "five": {
        "doc_comments": [],
        "doc_position": 0,
        "value_type": {
          "type": "PrimConstructor",
          "data": "Int"
        }
      },
      "get_a": {
        "doc_comments": [],
        "doc_position": 1,
        "value_type": {
          "type": "Function",
          "data": {
            "parameters": [
              {
                "type": "ConstructorAlias",
                "data": {
                  "constructor_kind": "Type",
                  "canonical_value": {
                    "module_name": [
                      null,
                      [
                        "Test"
                      ]
                    ],
                    "value": "Props"
                  },
                  "source_value": {
                    "module_name": null,
                    "value": "Props"
                  },
                  "alias_variables": [],
                  "aliased_type": {
                    "type": "RecordClosed",
                    "data": {
                      "kind": "Type",
                      "row": {
                        "a": {
                          "type": "PrimConstructor",
                          "data": "Int"
                        },
                        "b": {
                          "type": "PrimConstructor",
                          "data": "Bool"
                        },
                        "c": {
                          "type": "PrimConstructor",
                          "data": "Unit"
                        }
                      }
                    }
                  }
                }
              }
            ],
            "return_type": {
              "type": "PrimConstructor",
              "data": "Int"
            }
          }
        }
      }
    }
  },
  "types": {
    "Props": {
      "Alias": {
        "doc_comments": [],
        "type_name_span": {
          "start_offset": 37,
          "end_offset": 42
        },
        "kind": "Type",
        "aliased_type": {
          "type": "RecordClosed",
          "data": {
            "kind": "Type",
            "row": {
              "a": {
                "type": "PrimConstructor",
                "data": "Int"
              },
              "b": {
                "type": "PrimConstructor",
                "data": "Bool"
              },
              "c": {
                "type": "PrimConstructor",
                "data": "Unit"
              }
            }
          }
        },
        "alias_variables": []
      }
    }
  },
  "constructors": {},
  "values": {
    "get_a": {
      "doc_comments": [],
      "name_span": {
        "start_offset": 75,
        "end_offset": 80
      },
      "expression": {
        "expression": "Function",
        "data": {
          "span": {
            "start_offset": 83,
            "end_offset": 111
          },
          "function_type": {
            "type": "Function",
            "data": {
              "parameters": [
                {
                  "type": "ConstructorAlias",
                  "data": {
                    "constructor_kind": "Type",
                    "canonical_value": {
                      "module_name": [
                        null,
                        [
                          "Test"
                        ]
                      ],
                      "value": "Props"
                    },
                    "source_value": {
                      "module_name": null,
                      "value": "Props"
                    },
                    "alias_variables": [],
                    "aliased_type": {
                      "type": "RecordClosed",
                      "data": {
                        "kind": "Type",
                        "row": {
                          "a": {
                            "type": "PrimConstructor",
                            "data": "Int"
                          },
                          "b": {
                            "type": "PrimConstructor",
                            "data": "Bool"
                          },
                          "c": {
                            "type": "PrimConstructor",
                            "data": "Unit"
                          }
                        }
                      }
                    }
                  }
                }
              ],
              "return_type": {
                "type": "PrimConstructor",
                "data": "Int"
              }
            }
          },
          "binders": [
            [
              {
                "Variable": {
                  "span": {
                    "start_offset": 87,
                    "end_offset": 92
                  },
                  "name": "props"
                }
              },
              {
                "type": "ConstructorAlias",
                "data": {
                  "constructor_kind": "Type",
                  "canonical_value": {
                    "module_name": [
                      null,
                      [
                        "Test"
                      ]
                    ],
                    "value": "Props"
                  },
                  "source_value": {
                    "module_name": null,
                    "value": "Props"
                  },
                  "alias_variables": [],
                  "aliased_type": {
                    "type": "RecordClosed",
                    "data": {
                      "kind": "Type",
                      "row": {
                        "a": {
                          "type": "PrimConstructor",
                          "data": "Int"
                        },
                        "b": {
                          "type": "PrimConstructor",
                          "data": "Bool"
                        },
                        "c": {
                          "type": "PrimConstructor",
                          "data": "Unit"
                        }
                      }
                    }
                  }
                }
              }
            ]
          ],
          "body": {
            "expression": "RecordAccess",
            "data": {
              "span": {
                "start_offset": 104,
                "end_offset": 111
              },
              "field_type": {
                "type": "PrimConstructor",
                "data": "Int"
              },
              "target": {
                "expression": "LocalVariable",
                "data": {
                  "span": {
                    "start_offset": 104,
                    "end_offset": 109
                  },
                  "variable_type": {
                    "type": "ConstructorAlias",
                    "data": {
                      "constructor_kind": "Type",
                      "canonical_value": {
                        "module_name": [
                          null,
                          [
                            "Test"
                          ]
                        ],
                        "value": "Props"
                      },
                      "source_value": {
                        "module_name": null,
                        "value": "Props"
                      },
                      "alias_variables": [],
                      "aliased_type": {
                        "type": "RecordClosed",
                        "data": {
                          "kind": "Type",
                          "row": {
                            "a": {
                              "type": "PrimConstructor",
                              "data": "Int"
                            },
                            "b": {
                              "type": "PrimConstructor",
                              "data": "Bool"
                            },
                            "c": {
                              "type": "PrimConstructor",
                              "data": "Unit"
                            }
                          }
                        }
                      }
                    }
                  },
                  "variable": "props"
                }
              },
              "label": "a"
            }
          }
        }
      }
    },
    "five": {
      "doc_comments": [],
      "name_span": {
        "start_offset": 113,
        "end_offset": 117
      },
      "expression": {
        "expression": "Call",
        "data": {
          "span": {
            "start_offset": 120,
            "end_offset": 156
          },
          "call_type": {
            "type": "PrimConstructor",
            "data": "Int"
          },
          "function": {
            "expression": "LocalVariable",
            "data": {
              "span": {
                "start_offset": 120,
                "end_offset": 125
              },
              "variable_type": {
                "type": "Function",
                "data": {
                  "parameters": [
                    {
                      "type": "ConstructorAlias",
                      "data": {
                        "constructor_kind": "Type",
                        "canonical_value": {
                          "module_name": [
                            null,
                            [
                              "Test"
                            ]
                          ],
                          "value": "Props"
                        },
                        "source_value": {
                          "module_name": null,
                          "value": "Props"
                        },
                        "alias_variables": [],
                        "aliased_type": {
                          "type": "RecordClosed",
                          "data": {
                            "kind": "Type",
                            "row": {
                              "a": {
                                "type": "PrimConstructor",
                                "data": "Int"
                              },
                              "b": {
                                "type": "PrimConstructor",
                                "data": "Bool"
                              },
                              "c": {
                                "type": "PrimConstructor",
                                "data": "Unit"
                              }
                            }
                          }
                        }
                      }
                    }
                  ],
                  "return_type": {
                    "type": "PrimConstructor",
                    "data": "Int"
                  }
                }
              },
              "variable": "get_a"
            }
          },
          "arguments": [
            {
              "Expression": {
                "expression": "Record",
                "data": {
                  "span": {
                    "start_offset": 126,
                    "end_offset": 155
                  },
                  "record_type": {
                    "type": "ConstructorAlias",
                    "data": {
                      "constructor_kind": "Type",
                      "canonical_value": {
                        "module_name": [
                          null,
                          [
                            "Test"
                          ]
                        ],
                        "value": "Props"
                      },
                      "source_value": {
                        "module_name": null,
                        "value": "Props"
                      },
                      "alias_variables": [],
                      "aliased_type": {
                        "type": "RecordClosed",
                        "data": {
                          "kind": "Type",
                          "row": {
                            "a": {
                              "type": "PrimConstructor",
                              "data": "Int"
                            },
                            "b": {
                              "type": "PrimConstructor",
                              "data": "Bool"
                            },
                            "c": {
                              "type": "PrimConstructor",
                              "data": "Unit"
                            }
                          }
                        }
                      }
                    }
                  },
                  "fields": {
                    "a": {
                      "expression": "Int",
                      "data": {
                        "span": {
                          "start_offset": 132,
                          "end_offset": 133
                        },
                        "value": "5",
                        "value_type": {
                          "type": "PrimConstructor",
                          "data": "Int"
                        }
                      }
                    },
                    "b": {
                      "expression": "True",
                      "data": {
                        "span": {
                          "start_offset": 139,
                          "end_offset": 143
                        },
                        "value_type": {
                          "type": "PrimConstructor",
                          "data": "Bool"
                        }
                      }
                    },
                    "c": {
                      "expression": "Unit",
                      "data": {
                        "span": {
                          "start_offset": 149,
                          "end_offset": 153
                        },
                        "value_type": {
                          "type": "PrimConstructor",
                          "data": "Unit"
                        }
                      }
                    }
                  }
                }
              }
            }
          ]
        }
      }
    }
  },
  "values_toposort": [
    "get_a",
    "five"
  ]
}